src/query/plan/rewrite/periodic_delete.hpp

```diff
@@ -0,0 +1,79 @@
+// Copyright 2024 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <memory>
+#include "query/plan/operator.hpp"
+
+namespace memgraph::query::plan {
+
+namespace impl {
+
+template <class TDbAccessor>
+class PeriodicDeleteRewriter final : public HierarchicalLogicalOperatorVisitor {
+ public:
+  PeriodicDeleteRewriter(SymbolTable *symbolTable, AstStorage *astStorage, TDbAccessor *db)
+      : symbol_table(symbolTable), ast_storage(astStorage), db(db) {}
+
+  ~PeriodicDeleteRewriter() override = default;
+
+  using HierarchicalLogicalOperatorVisitor::PostVisit;
+  using HierarchicalLogicalOperatorVisitor::PreVisit;
+  using HierarchicalLogicalOperatorVisitor::Visit;
+
+  bool Visit(Once &t) override { return true; }
+
+  bool PreVisit(PeriodicCommit &op) override {
+    commit_frequency_ = op.commit_frequency_;
+    return true;
+  }
+
+  bool PreVisit(PeriodicSubquery &op) override {
+    periodic_subquery_seen_ = true;
+    return true;
+  }
+
+  bool PreVisit(Delete &op) override {
+    // since periodic commit needs to flush deltas over time, buffer size needs to be put in delete
+    // so it can know when to trigger delete of nodes and/or relationships
+    if (commit_frequency_) {
+      op.buffer_size_ = commit_frequency_;
+    }
```

Comment:
I had to check query planning for this part but let me know if I got something wrong. I am not sure why will `prerevisit` Delete after `previsiting` PeriodicCommit but since we are having periodic commit always at the end of our PullPlan (because we generate it at GenReturnBody function, which is always after HandleWriteClause in rule_based_planner.hpp), this means we will previsit in our visitor first periodic commit and then Delete, which means this should always work as it should (meaning if we actually have periodic commit + delete, we will notice it)

Reply:
Plan's first operator is always Produce or Empty result (the last part of the query). Meaning, PeriodicCommit's previsit needs to be before the Delete, in order for periodic commit to influence delete and transfer the commit frequency, i.e. buffer size to delete.

Reply:
Yes yes, that is what I wanted to confirm, exactly. So `PeriodicCommit` will be visited in `previsit` function before `Delete` every time, so this should work